---
title: "scMuffin"
author:
- name: Noemi Di Nanni*
  affiliation: CNR-ITB
- name: Valentina Nale*
  affiliation: CNR-ITB
- name: Alice Chiodi
  affiliation: CNR-ITB
- name: Cinzia Cocola
  affiliation: CNR-ITB
- name: Ingrid Cifola
  affiliation: CNR-ITB
- name: Eleonora Piscitelli
  affiliation: CNR-ITB
- name: Ileana Zucchi
  affiliation: CNR-ITB
- name: Rolland Reinbold
  affiliation: CNR-ITB
- name: Luciano Milanesi
  affiliation: CNR-ITB
- name: Alessandra Mezzelani
  affiliation: CNR-ITB
- name: Paride Pelucchi*
  affiliation: CNR-ITB
- name: Ettore Mosca*
  affiliation: CNR-ITB
  email: ettore.mosca@itb.cnr.it
package: scMuffin
output:
  BiocStyle::html_document:
    toc: true
abstract: >
  INTRODUCTION: Single cell (SC) analysis is crucial to study the complex cellular heterogeneity of solid tumors, which is one of the main obstacles for the development of effective cancer treatments. Such tumors typically contain a mixture of cells with aberrant genomic and expression profiles affecting specific sub-populations that have a pivotal role in cancer progression, whose identification eludes bulk approaches. We present a MUlti-Features INtegrative approach for SC data analysis (scMuffin) that characterizes cell identity on the basis of multiple and complementary criteria. MATERIALS AND METHODS: Cell markers sources: CSEA, PanglaoDB. Pathways sources: NCBI Biosystems, MSigDB. Gene set expression is assessed by a fast algorithm that uses comparable control-gene sets. CNVs are estimated using adjacent genes. Lineage analysis is computed by Monocle; multipotency is assessed by LandScent. The association between the various features and cell clusters is assessed by chi-squared and enrichment-based approaches. RESULTS: scMuffin provides functions to calculate a series of qualitative and quantitative scores, such as: expression of markers for normal and tumor conditions, pathway activity, cell hierarchy, multipotency state, copy number variations and cell cycle state. Cell-level scores are used for cell cluster annotation and combined to obtain alternative cell clusters. scMuffin integrates any type of cell- or cluster-associated data, and can be used for single-cell multi-omics analyses (e.g. mutations, gene expression). As a proof-of-principle, we studied a public dataset of human gliomas. CONCLUSIONS: scMuffin combines several tools to shed light on the identity of tumors cells and spot subtle cell types. The package is available for non-commercial users at the URL https://github.com/emosca-cnr/scMuffin.
vignette: >
  %\VignetteIndexEntry{scMuffin}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

# Getting started

```{r, include=TRUE, echo=FALSE, out.width="60%", fig.align = "center", fig.cap="scMuffing logo"}
knitr::include_graphics("images/scMuffin_logo_pontillism.jpg")
```

```{r, include=TRUE, eval=FALSE}
library(devtools)
install_github("emosca-cnr/scMuffin", build_vignettes = TRUE)
library(scMuffin)
```

# Cell-level quantities

## Cell-level estimation of gene set expression

scMuffin provides functions to set up one or more collections of gene sets and perform cell-level estimation of gene set expression in relation to an empirical null model. This can be applied to any gene set, and can therefore be used to estimate various cell’s phenotypes, like pathway activities or marker set expression. It requires in input a normalized genes-by-cells expression matrix. In the following example, we estimate gene sets score for the CancerSea gene sets using the Seurat object "seu_obj":

```{r, include=TRUE, eval=FALSE}
gsc <- prepare_signatures(gsc="SIG_CancerSEA", genes = rownames(seu_obj), genes_min = 3)
res_gss <- calculate_gs_scores(genes_by_cells = GetAssayData(seu_obj), gs_list = gsc$SIG_CancerSEA)
```

Cell cluster average gene set score can be quantified using calculate_gs_scores_in_clusters:
```{r, include=TRUE, eval=FALSE}
res_gss_cls <- calculate_gs_scores_in_clusters(gs_scores_obj = res_gss, cell_clusters = seu_obj$seurat_clusters)
```

## CNV inference

The function calculate_CNV basically retrieves the genomic locations and performs the CNV estimation; cluster_by_features calculates the clustering; apply_CNV_reference redefines the CNV levels on the basis of (optional) reference cells; the dedicated plotting function CNV_heatmap handles the visualization, where the cluster of cells that contains the reference is marked. Here’s an example that illustrates CNV inference using a 100 genes window size and a reference profile from the The Genotype-Tissue Expression project (GTEx) portal (13): 

```{r, include=TRUE, eval=FALSE}
cnv_res <- calculate_CNV(genes_by_cells = as.matrix(GetAssayData(seu_obj)), wnd_size = 100, reference = GTEx_mean, mc.cores = 2, )
cnv_clustering <- cluster_by_features(cnv_res, cnv=TRUE)
cnv_res_ref <- apply_CNV_reference(cnv = cnv_res, cnv_clustering = cnv_clustering, reference="reference")
cnv_res_ref <- heatmap_CNV(cnv = cnv_res, cnv_clustering = cnv_clustering, ref_cluster = "reference")
```
The calculation required approximately 10 minutes on 2 cores.

## The global state of the chromatin
The chromatin state can be calculated using genes-by-cells count matrix, as follows:

```{r, include=TRUE, eval=FALSE}
res_chrom_state <- exp_rate(GetAssayData(seu_obj, slot = "count"), min_counts = 5)
```

where 5 is the required threshold above which a gene is considered expressed.

## Cell  proliferation

Cell proliferation is quantified considering the maximum value of the two gene sets scores for G1/S and G2/M genes:
```{r, include=TRUE, eval=FALSE}
res_prol <- proliferation_analysis(GetAssayData(seu_obj))
```

## Trajectory analysis
Diffusion maps identify differentiation trajectories. scMuffin relies on the diffusion pseudo time calculation available in the R package "destiny". Here we calculate the diffusion map over the first 50 PC (to speed up computation), using a random cell to obtain diffusion pseudotimes:

```{r, include=TRUE, eval=FALSE}
res_dm <- diff_map(GetAssayData(seu_obj), root_cell = "random", n_pcs=50)
```

# Clustering and Annotation

## Assembling features and clusterings

Definition of the object that containts the features:
```{r, include=TRUE, eval=FALSE}
feat_obj <- create_features(t(res_signatures$SIG_CSEA$signatures_by_cells))
feat_obj <- add_features(feat_obj, as.data.frame(res_chrom_state))
```

Clustering by features:
```{r, include=TRUE, eval=FALSE}
feat_clust <- cluster_by_features(feat_obj)
```

Definition of the object that contains the clusterings:
```{r, include=TRUE, eval=FALSE}
clust_obj <- create_clusterings(cell_id = rownames(seu_obj), values = data.frame(global=seu_obj@meta.data$seurat_clusters))
clust_obj <- add_clusterings(clust_obj, cell_id = names(clust_by_csea$clusters), values = data.frame(cancer_state=clust_by_csea$clusters)))
```

## Assessing clustering enrichment
The enrichment of each cluster of cells in terms of quantitative or categorical features can be assessed using:
```{r, include=TRUE, eval=FALSE}
cl_enrich <- assess_cluster_enrichment(features = features, clusterings = clusterings)
cl_enrich_table <- extract_cluster_enrichment_table(cl_enrich, q_type = "nes", c_type = "er")
```

## Two-samples comparison
To compare the clusterings of two samples on the basis of the expression of their markers:
```{r, include=TRUE, eval=FALSE}
sim_res <- quantify_samples_similarity(gbc_1, gbc_2, clusters_1, clusters_2, cluster_markers_1, cluster_markers_2)
```

## Comparison of clusterings and definition of meta-clusters

```{r, include=TRUE, eval=FALSE}
cl_list <- clusterings_to_list(clusterings[, c("global", "GB500", "ribosomes")])
ov_mat <- overlap_matrix(cl_list)
meta_cl <- meta_cluster(ov_mat, n_step = 10, max_clust = 15)
meta_cl_cell <- get_meta_clusters(cl_list, meta_cl)
```

# Visualization

CNV heatmap
```{r, include=TRUE, eval=FALSE}
cnv_res_ref <- CNV_heatmap(cnv = cnv_res, cnv_clustering = cnv_clustering, reference="reference")
```

Boxplot of cluster enrichment in a quantitative feature:
```{r, include=TRUE, eval=FALSE}
top_feat_lab <- boxplot_cluster(feat_obj, cell_clusters = , cluster_enrichment = clust_enrich$global, fdr_threshold = 0.05, only_pos_nes = TRUE)
```

Boxplot of cluster enrichment in a categorical feature:
```{r, include=TRUE, eval=FALSE}
top_feat_lab <- dotplot_cluster(feat_obj, cell_clusters = , cluster_enrichment = clust_enrich$global, fdr_threshold = 0.05, only_pos_nes = TRUE)
```

Plot umap labelling clusters using the enriched scores:
```{r, include=TRUE, eval=FALSE}
plot_umap(seu_obj, group.by = "seurat_clusters", labels=lapply(top_feat_lab, function(x) x[1:2]), lab_size=2)
```

Features-by-clusters heatmap:
```{r, include=TRUE, eval=FALSE}
plot_heatmap_features_by_clusters(list(feature_markers=t(clust_enrich$global$nes)), ntop = ncol(clust_enrich$global$nes))
```

plot meta clusters:
```{r, include=TRUE, eval=FALSE}
plot_meta_clusters(ov_mat, meta_cl)
```
