<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scMuffin">
<title>scMuffin package vignettes • scMuffin</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="scMuffin package vignettes">
<meta property="og:description" content="scMuffin">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scMuffin</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/scMuffin.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>scMuffin package vignettes</h1>
            
      
      
      <div class="d-none name"><code>scMuffin.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p><img src="images/logo.jpg" width="300"></p>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>scMuffin requires R &gt;= 4.0.0, due to some of its dependencies, like Seurat <span class="citation">(Hao et al. 2021)</span>. R can be installed from CRAN at the URL <a href="https://cran.r-project.org/index.html" class="external-link uri">https://cran.r-project.org/index.html</a>.</p>
<p>To succesfully install scMuffin you need some packages from Bioconductor (<a href="https://bioconductor.org" class="external-link uri">https://bioconductor.org</a>) and github (<a href="https://github.com/" class="external-link uri">https://github.com/</a>). These packages can be installed using the following commands:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BiocStyle"</span>, <span class="st">"ComplexHeatmap"</span>, <span class="st">"DESeq2"</span>, <span class="st">"org.Hs.eg.db"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"theislab/destiny"</span><span class="op">)</span></span></code></pre></div>
<p>The other dependencies, if missing, should be automatically installed using the following command:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"emosca-cnr/scMuffin"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>To load the package:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scMuffin</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="input">Input<a class="anchor" aria-label="anchor" href="#input"></a>
</h3>
<p>scMuffin is intended to be used downstream general purpose tasks like quality control, normalization, cell clustering and dataset integration, for which there are dedicated tools, such as Seurat <span class="citation">(Hao et al. 2021)</span>. scMuffin requires three inputs:</p>
<ul>
<li>genes-by-cells raw counts matrix;</li>
<li>genes-by-cells normalized expression matrix;</li>
<li>a partition of cells (cell clusters).</li>
</ul>
<p>Typically, the two genes-by-cells matrices have already been filtered to exclude low quality cells and genes that could negatively affect the analyses. However, the characterization of cells that can be achieved with scMuffin offers insights that can be used to (further) filter a dataset and/or to decide on which cells apply particular analyses (e.g. biomarker identification). In general, according to research questions and experimental design one may want to apply strong or mild filters before using scMuffin.</p>
</div>
<div class="section level3">
<h3 id="obtaining-the-data-used-in-this-vignette">Obtaining the data used in this vignette<a class="anchor" aria-label="anchor" href="#obtaining-the-data-used-in-this-vignette"></a>
</h3>
<p>To run the examples included in this vignette you need the following data:</p>
<ul>
<li>two single cell datasets with counts as well as normalized expression values;</li>
<li>cell clusters;</li>
<li>UMAP visualization;</li>
<li>cluster markers.</li>
</ul>
<p>Here’s an example of how these data can be obtained from a publicly available study on High-Grade Glioma <span class="citation">(J. Yuan et al. 2018)</span> (GSE103224):</p>
<ol style="list-style-type: decimal">
<li>download the filtered genes-by-cells mtrices from the URLs:</li>
</ol>
<ul>
<li>sample PJ016: <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM2758471" class="external-link uri">https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM2758471</a>;</li>
<li>sample PJ017: <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM2758472" class="external-link uri">https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM2758472</a>.</li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li>read the text files into R and use the Seurat package <span class="citation">(Hao et al. 2021)</span> to perform a series of typical tasks (e.g. normalization, clustering); here’s a concise example using sample PJ016:</li>
</ol>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gbc_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"GSM2758471_PJ016.filtered.matrix.txt"</span><span class="op">)</span></span>
<span><span class="va">sym</span> <span class="op">&lt;-</span> <span class="va">gbc_counts</span><span class="op">[</span>, <span class="st">"V2"</span><span class="op">]</span></span>
<span><span class="va">gbc_counts</span><span class="op">[</span>, <span class="st">"V2"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">gbc_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/keep_strongest_representative.html">keep_strongest_representative</a></span><span class="op">(</span><span class="va">gbc_counts</span>, <span class="va">sym</span><span class="op">)</span></span>
<span><span class="va">seu_obj_1</span> <span class="op">&lt;-</span> <span class="fu">CreateSeuratObject</span><span class="op">(</span>count<span class="op">=</span><span class="va">gbc_counts</span>, project<span class="op">=</span><span class="st">"PJ016"</span><span class="op">)</span></span>
<span><span class="va">seu_obj_1</span> <span class="op">&lt;-</span> <span class="fu">NormalizeData</span><span class="op">(</span><span class="va">seu_obj_1</span><span class="op">)</span></span>
<span><span class="va">seu_obj_1</span> <span class="op">&lt;-</span> <span class="fu">FindVariableFeatures</span><span class="op">(</span><span class="va">seu_obj_1</span><span class="op">)</span></span>
<span><span class="va">seu_obj_1</span> <span class="op">&lt;-</span> <span class="fu">ScaleData</span><span class="op">(</span><span class="va">seu_obj_1</span><span class="op">)</span></span>
<span><span class="va">seu_obj_1</span> <span class="op">&lt;-</span> <span class="fu">RunPCA</span><span class="op">(</span><span class="va">seu_obj_1</span><span class="op">)</span></span>
<span><span class="va">seu_obj_1</span> <span class="op">&lt;-</span> <span class="fu">FindNeighbors</span><span class="op">(</span><span class="va">seu_obj_1</span><span class="op">)</span></span>
<span><span class="va">seu_obj_1</span> <span class="op">&lt;-</span> <span class="fu">FindClusters</span><span class="op">(</span><span class="va">seu_obj_1</span><span class="op">)</span></span>
<span><span class="va">seu_obj_1</span> <span class="op">&lt;-</span> <span class="fu">RunUMAP</span><span class="op">(</span><span class="va">seu_obj_1</span><span class="op">)</span></span></code></pre></div>
<p>The genes-by-cells count matrix available at NCBI GEO is provided with Ensembl identifiers and gene symbols, which are not unique. So, after having read the table, we take advantage of the function <code>keep_strongest_representative</code>, which defines a genes-by-cells matrix with symbols as row names. In case of multiple Ensembl ids mapped to the same symbol, the row with the highest average count is kept as representative of the gene. In a real scenario, the functions listed above should be run using appropriate parameters that take into account the specificities of the dataset under consideration and the objectives of the analysis that one wants to perform. Here we just wanted to provide a means to obtain a dataset to follow this vignette. Please read the Seurat documentation for further details about the functions listed above.</p>
</div>
</div>
<div class="section level2">
<h2 id="the-scmuffinlist-data-structure">The scMuffinList data structure<a class="anchor" aria-label="anchor" href="#the-scmuffinlist-data-structure"></a>
</h2>
<p>All data and results of scMuffin are stored in the scMuffinList, so that every function takes in input such structure and returns a modified version of it. The scMuffinList can be created as follows:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMuffinList_demo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_scMuffinList.html">create_scMuffinList</a></span><span class="op">(</span>counts<span class="op">=</span><span class="fu">GetAssayData</span><span class="op">(</span><span class="va">seu_obj_1</span>, assay<span class="op">=</span><span class="st">"RNA"</span>, slot<span class="op">=</span><span class="st">"counts"</span><span class="op">)</span>, normalized<span class="op">=</span><span class="fu">GetAssayData</span><span class="op">(</span><span class="va">seu_obj_1</span>, assay<span class="op">=</span><span class="st">"RNA"</span>, slot<span class="op">=</span><span class="st">"data"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Many analyses require cell clusters. Any partition can be added as follows:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMuffinList_demo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_partitions.html">add_partitions</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span>, clusters <span class="op">=</span> <span class="va">seu_obj_1</span><span class="op">$</span><span class="va">seurat_clusters</span>, partition_id <span class="op">=</span> <span class="st">"global_expr"</span><span class="op">)</span></span></code></pre></div>
<p>where seu_obj_1$seurat_clusters is a named vector that contains cells clusters and whose names are cell labels.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span><span class="op">$</span><span class="va">partitions</span><span class="op">[</span>, <span class="fl">2</span>, drop<span class="op">=</span><span class="cn">F</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    CNV</span></span>
<span><span class="co">## V3   0</span></span>
<span><span class="co">## V4   0</span></span>
<span><span class="co">## V5   2</span></span>
<span><span class="co">## V6   0</span></span>
<span><span class="co">## V7   2</span></span>
<span><span class="co">## V8   1</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="gene-set-scoring">Gene set scoring<a class="anchor" aria-label="anchor" href="#gene-set-scoring"></a>
</h2>
<p>scMuffin provides functions to set up one or more collections of gene sets and perform cell-level estimation of gene set expression in relation to an empirical null model. This can be applied to any gene set, and can therefore be used to estimate various cell’s phenotypes, like pathway activities or marker set expression.</p>
<div class="section level3">
<h3 id="assembling-input-gene-sets">Assembling input gene sets<a class="anchor" aria-label="anchor" href="#assembling-input-gene-sets"></a>
</h3>
<p>The function <code>prepare_gsls</code> retrieves gene sets from CellMarker [Zhang2019], PanglaoDB <span class="citation">(Franzén, Gan, and Björkegren 2019)</span>, CancerSEA <span class="citation">(H. Yuan et al. 2019)</span> and MSigDB <span class="citation">(Subramanian et al. 2005)</span>, and accepts custom gene set as well. The full list of gene sets available whitin CellMarker, PanglaoDB and CancerSEA collections can be listed using:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"gsls_EntrezID"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"gsls_Symbol"</span><span class="op">)</span></span></code></pre></div>
<p>while in the case of MSigDB we can use its dedicated functions:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">msigdbr</span><span class="fu">::</span><span class="fu"><a href="https://igordot.github.io/msigdbr/reference/msigdbr_collections.html" class="external-link">msigdbr_collections</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">msigdbr</span><span class="fu">::</span><span class="fu"><a href="https://igordot.github.io/msigdbr/reference/msigdbr_species.html" class="external-link">msigdbr_species</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>The gene sets of interest can be selected acting on the corresponding arguments <code>CM_tissues</code>, <code>PNDB_tissues</code>, and <code>msigdb_hs_cat_subcat</code>. In the case of CellMarker and PanglaoDB we can specify a list of desired tissues. The full set can be listed by means of show_tissues. Here we show just the first part of the output:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/show_tissues.html">show_tissues</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Here is an example:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gsc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_gsls.html">prepare_gsls</a></span><span class="op">(</span>gs_sources <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CancerSEA"</span>, <span class="st">"PNDB"</span><span class="op">)</span>, PNDB_tissues <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Brain"</span><span class="op">)</span>, scMuffinList <span class="op">=</span> <span class="va">scMuffinList_demo</span>, genes_min <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p>Note that in the case of MSigDB we have to set up a <code>data.frame</code> to specify species, category and sub_category of the gene set collections.</p>
</div>
<div class="section level3">
<h3 id="calculate-gene-set-scores-at-cell-and-cluster-level">Calculate Gene set scores at cell and cluster level<a class="anchor" aria-label="anchor" href="#calculate-gene-set-scores-at-cell-and-cluster-level"></a>
</h3>
<p>In the following example, we estimate the scores for CancerSEA gene sets:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMuffinList_demo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_gs_scores.html">calculate_gs_scores</a></span><span class="op">(</span>scMuffinList <span class="op">=</span> <span class="va">scMuffinList_demo</span>, gs_list <span class="op">=</span> <span class="va">gsc</span><span class="op">$</span><span class="va">CancerSEA</span><span class="op">)</span></span></code></pre></div>
<p>The results of any analysis are stored as elements of the <code>scMuffinList</code>. The gene set scoring engine stores its results in <code>scMuffinList$gene_set_scoring</code>, where <code>summary</code> is a cells-by-features data.frame with cell scores, and <code>full</code> contains a series of additional details for every gene set:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span><span class="op">$</span><span class="va">gene_set_scoring</span><span class="op">$</span><span class="va">summary</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    CSEA_Angiogenesis CSEA_Apoptosis CSEA_Cell_Cycle CSEA_Differentiation</span></span>
<span><span class="co">## V3       -0.08918562     0.04322924      0.17300016            0.2360843</span></span>
<span><span class="co">## V4       -0.01725770     0.10940020      0.28114799            0.2767576</span></span>
<span><span class="co">## V5        0.11402356     0.05401112     -0.02053489            0.1757745</span></span>
<span><span class="co">## V6       -0.06014349     0.02707061      0.37283078            0.2492808</span></span>
<span><span class="co">## V7        0.10562932     0.24878223     -0.08555433            0.1855214</span></span>
<span><span class="co">## V8        0.16643327     0.27654624     -0.08090808            0.1075169</span></span>
<span><span class="co">##    CSEA_DNA_damage</span></span>
<span><span class="co">## V3      0.13994803</span></span>
<span><span class="co">## V4      0.31254961</span></span>
<span><span class="co">## V5      0.16894379</span></span>
<span><span class="co">## V6      0.18368743</span></span>
<span><span class="co">## V7      0.03395132</span></span>
<span><span class="co">## V8      0.11673244</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span><span class="op">$</span><span class="va">gene_set_scoring</span><span class="op">$</span><span class="va">full</span><span class="op">$</span><span class="va">CSEA_Angiogenesis</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         case case.N case.AV nmark_min avg_control control.AV null_ok</span></span>
<span><span class="co">## V3 0.5958431     71      22      TRUE   0.6850288         99    TRUE</span></span>
<span><span class="co">## V4 0.6793401     71      22      TRUE   0.6965978        100    TRUE</span></span>
<span><span class="co">## V5 0.8097307     71      24      TRUE   0.6957071         99    TRUE</span></span>
<span><span class="co">## V6 0.6537237     71      16      TRUE   0.7138672         99    TRUE</span></span>
<span><span class="co">## V7 0.8049070     71      24      TRUE   0.6992777         99    TRUE</span></span>
<span><span class="co">## V8 0.8873211     71      19      TRUE   0.7208879        100    TRUE</span></span>
<span><span class="co">##    avg_delta_score delta_score</span></span>
<span><span class="co">## V3     -0.08918562 -0.08918562</span></span>
<span><span class="co">## V4     -0.01725770 -0.01725770</span></span>
<span><span class="co">## V5      0.11402356  0.11402356</span></span>
<span><span class="co">## V6     -0.06014349 -0.06014349</span></span>
<span><span class="co">## V7      0.10562932  0.10562932</span></span>
<span><span class="co">## V8      0.16643327  0.16643327</span></span></code></pre>
<p>The function <code>calculate_gs_scores_in_clusters</code> defines the median values of gene set scores in every cluster of a given partition id:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMuffinList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_gs_scores_in_clusters.html">calculate_gs_scores_in_clusters</a></span><span class="op">(</span><span class="va">scMuffinList</span>, partition_id <span class="op">=</span> <span class="st">"global_expr"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMuffinList_demo</span><span class="op">$</span><span class="va">cluster_data</span><span class="op">$</span><span class="va">global_expr</span><span class="op">$</span><span class="va">gene_set_scoring</span><span class="op">$</span><span class="va">summary</span></span></code></pre></div>
<pre><code><span><span class="co">##   CSEA_Angiogenesis CSEA_Apoptosis CSEA_Cell_Cycle CSEA_Differentiation</span></span>
<span><span class="co">## 0        0.03466622     0.04455260    -0.027667196            0.1546205</span></span>
<span><span class="co">## 1        0.06283949    -0.05569218     0.119220367            0.1789833</span></span>
<span><span class="co">## 2        0.02561436     0.06234254    -0.011424615            0.1577567</span></span>
<span><span class="co">## 3        0.01546160     0.05865145     0.002646498            0.1788135</span></span>
<span><span class="co">## 4        0.01048663     0.03058474     0.002924762            0.1105544</span></span>
<span><span class="co">## 5        0.15187334     0.05179504    -0.001454732            0.2022856</span></span>
<span><span class="co">## 6        0.04492144    -0.02827166     0.037926976            0.1453372</span></span>
<span><span class="co">## 7        0.02152934    -0.06650877     0.298973763            0.2397246</span></span>
<span><span class="co">## 8        0.06987440     0.04130317    -0.035846983            0.1945336</span></span>
<span><span class="co">##   CSEA_DNA_damage CSEA_DNA_repair   CSEA_EMT CSEA_Hypoxia CSEA_Inflammation</span></span>
<span><span class="co">## 0      0.06477004    -0.010409047 0.24631020  0.161630871      -0.001521447</span></span>
<span><span class="co">## 1      0.14481900     0.151772082 0.10215837  0.021980962      -0.017672214</span></span>
<span><span class="co">## 2      0.16351859    -0.013112346 0.24304692  0.286081417       0.151495357</span></span>
<span><span class="co">## 3      0.15417556     0.002515808 0.18619531  0.295312377       0.112050700</span></span>
<span><span class="co">## 4      0.13590319     0.017085982 0.24579872  0.316156629       0.087552771</span></span>
<span><span class="co">## 5      0.16387229     0.008464217 0.32835167  0.284878014       0.103306520</span></span>
<span><span class="co">## 6      0.21480019     0.083121431 0.11631038  0.018344308       0.062785742</span></span>
<span><span class="co">## 7      0.14811608     0.143265241 0.07557108  0.008915609       0.020922322</span></span>
<span><span class="co">## 8      0.25436327     0.067705312 0.14146636 -0.020107706       0.027748835</span></span>
<span><span class="co">##   CSEA_Invasion CSEA_Metastasis CSEA_Proliferation CSEA_Quiescence</span></span>
<span><span class="co">## 0     0.1491915       0.3167512        -0.03419802     -0.04874756</span></span>
<span><span class="co">## 1     0.4192880       0.1410125         0.16986820     -0.01040003</span></span>
<span><span class="co">## 2     0.1947080       0.3166050        -0.02744911      0.09093454</span></span>
<span><span class="co">## 3     0.2183125       0.3104895         0.01500294      0.03456636</span></span>
<span><span class="co">## 4     0.1516411       0.2714825         0.08584974      0.14455138</span></span>
<span><span class="co">## 5     0.2210099       0.3680268        -0.01962822      0.05680748</span></span>
<span><span class="co">## 6     0.2105844       0.1090950         0.03454176      0.05277640</span></span>
<span><span class="co">## 7     0.5083292       0.1830839         0.35488900     -0.04732778</span></span>
<span><span class="co">## 8     0.2925115       0.1651046         0.14961468      0.13518170</span></span>
<span><span class="co">##   CSEA_Stemness</span></span>
<span><span class="co">## 0     0.2592561</span></span>
<span><span class="co">## 1     0.3960544</span></span>
<span><span class="co">## 2     0.2616661</span></span>
<span><span class="co">## 3     0.2791968</span></span>
<span><span class="co">## 4     0.3027992</span></span>
<span><span class="co">## 5     0.2357361</span></span>
<span><span class="co">## 6     0.3783765</span></span>
<span><span class="co">## 7     0.4564071</span></span>
<span><span class="co">## 8     0.3973588</span></span></code></pre>
<p>These mean values are useful to obtain a concise visualization of gene set expression throughout the dataset, using the function <code>plot_heatmap_features_by_clusters</code>:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_heatmap_features_by_clusters.html">plot_heatmap_features_by_clusters</a></span><span class="op">(</span><span class="va">scMuffinList</span>, feature_source <span class="op">=</span> <span class="st">"gss"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="images/gene_set_scores_clusters.png" width="600" alt=""><p class="caption">caption</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="cnv-inference">CNV inference<a class="anchor" aria-label="anchor" href="#cnv-inference"></a>
</h2>
<div class="section level3">
<h3 id="cnv-calculation">CNV calculation<a class="anchor" aria-label="anchor" href="#cnv-calculation"></a>
</h3>
<p>CNV inference is performed by the function <code>CNV_analysis</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMuffinList_demo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CNV_analysis.html">CNV_analysis</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span><span class="op">)</span></span></code></pre></div>
<p>The results are stored in <code>scMuffinList_demo$CNV</code>, where summary contains the “CNV_signal” and <code>scMuffinList_demo$CNV$full$CNV</code> the regions-by-cells CNV matrix; the element <code>scMuffinList_demo$CNV$full$regions2genes</code> is important to map the original data into the CNV regions; lastly, the <code>scMuffinList_demo$CNV$full$detected_cnv_regions</code> is a data.frame that list the CNV regions detected in each chromosome and cell cluster:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span><span class="op">$</span><span class="va">CNV</span><span class="op">$</span><span class="va">summary</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    CNV_signal</span></span>
<span><span class="co">## V3   41.61357</span></span>
<span><span class="co">## V4   48.08947</span></span>
<span><span class="co">## V5   39.13876</span></span>
<span><span class="co">## V6   44.46613</span></span>
<span><span class="co">## V7   36.90114</span></span>
<span><span class="co">## V8   44.93450</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span><span class="op">$</span><span class="va">CNV</span><span class="op">$</span><span class="va">full</span><span class="op">$</span><span class="va">CNV</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                                    V3         V4          V5</span></span>
<span><span class="co">## chr1__LINC00115_826205__UQCRHL_15807161   -0.09439528 -0.1298259 -0.08592605</span></span>
<span><span class="co">## chr1__LINC01128_827797__FLJ37453_15834214 -0.09439528 -0.1298259 -0.08592605</span></span>
<span><span class="co">## chr1__SAMD11_923922__SPEN_15847706        -0.09759899 -0.1330395 -0.08592605</span></span>
<span><span class="co">## chr1__NOC2L_944202__ZBTB17_15941868       -0.09759899 -0.1254592 -0.08592605</span></span>
<span><span class="co">## chr1__HES4_998963__FBXO42_16246839        -0.10080269 -0.1198235 -0.08592605</span></span>
<span><span class="co">## chr1__ISG15_1013496__SZRD1_16367241       -0.10080269 -0.1174013 -0.08592605</span></span>
<span><span class="co">##                                                    V6          V7         V8</span></span>
<span><span class="co">## chr1__LINC00115_826205__UQCRHL_15807161   -0.07809396 -0.09126164 -0.1287080</span></span>
<span><span class="co">## chr1__LINC01128_827797__FLJ37453_15834214 -0.07809396 -0.08801308 -0.1287080</span></span>
<span><span class="co">## chr1__SAMD11_923922__SPEN_15847706        -0.07809396 -0.08801308 -0.1229857</span></span>
<span><span class="co">## chr1__NOC2L_944202__ZBTB17_15941868       -0.07485419 -0.08801308 -0.1197168</span></span>
<span><span class="co">## chr1__HES4_998963__FBXO42_16246839        -0.08053094 -0.09126164 -0.1164479</span></span>
<span><span class="co">## chr1__ISG15_1013496__SZRD1_16367241       -0.07729117 -0.09126164 -0.1131790</span></span>
<span><span class="co">##                                                   V9        V10        V11</span></span>
<span><span class="co">## chr1__LINC00115_826205__UQCRHL_15807161   -0.1338054 -0.1115354 0.03965468</span></span>
<span><span class="co">## chr1__LINC01128_827797__FLJ37453_15834214 -0.1370741 -0.1115354 0.04297391</span></span>
<span><span class="co">## chr1__SAMD11_923922__SPEN_15847706        -0.1338054 -0.1082252 0.04877464</span></span>
<span><span class="co">## chr1__NOC2L_944202__ZBTB17_15941868       -0.1338054 -0.1115354 0.05457536</span></span>
<span><span class="co">## chr1__HES4_998963__FBXO42_16246839        -0.1338054 -0.1115354 0.05705685</span></span>
<span><span class="co">## chr1__ISG15_1013496__SZRD1_16367241       -0.1338054 -0.1082252 0.06037609</span></span>
<span><span class="co">##                                                   V12</span></span>
<span><span class="co">## chr1__LINC00115_826205__UQCRHL_15807161   -0.08178787</span></span>
<span><span class="co">## chr1__LINC01128_827797__FLJ37453_15834214 -0.08178787</span></span>
<span><span class="co">## chr1__SAMD11_923922__SPEN_15847706        -0.08178787</span></span>
<span><span class="co">## chr1__NOC2L_944202__ZBTB17_15941868       -0.07842147</span></span>
<span><span class="co">## chr1__HES4_998963__FBXO42_16246839        -0.07842147</span></span>
<span><span class="co">## chr1__ISG15_1013496__SZRD1_16367241       -0.07254742</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span><span class="op">$</span><span class="va">CNV</span><span class="op">$</span><span class="va">full</span><span class="op">$</span><span class="va">regions2genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $chr1__LINC00115_826205__UQCRHL_15807161</span></span>
<span><span class="co">##       symbol location</span></span>
<span><span class="co">## 11 LINC00115   826205</span></span>
<span><span class="co">## 12 LINC01128   827797</span></span>
<span><span class="co">## 13    SAMD11   923922</span></span>
<span><span class="co">## 14     NOC2L   944202</span></span>
<span><span class="co">## 15      HES4   998963</span></span>
<span><span class="co">## 16     ISG15  1013496</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span><span class="op">$</span><span class="va">CNV</span><span class="op">$</span><span class="va">full</span><span class="op">$</span><span class="va">detected_cnv_regions</span><span class="op">$</span><span class="va">chr1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    chr                                   start start.loc</span></span>
<span><span class="co">## 1 chr1 chr1__LINC00115_826205__UQCRHL_15807161    826205</span></span>
<span><span class="co">## 2 chr1    chr1__PHF13_6613730__LYPLA2_23791144   6613730</span></span>
<span><span class="co">## 3 chr1    chr1__CAMTA1_6785323__FUCA1_23845080   6785323</span></span>
<span><span class="co">## 4 chr1  chr1__HIPK1_113953876__CHTOP_153633981 113953876</span></span>
<span><span class="co">## 5 chr1   chr1__CD58_116514533__RAB13_153981604 116514533</span></span>
<span><span class="co">## 6 chr1  chr1__S100A2_153561108__DEDD_161120977 153561108</span></span>
<span><span class="co">##                                      stop  stop.loc cluster   length</span></span>
<span><span class="co">## 1   chr1__KLHL21_6590723__PITHD1_23778417  23778417       0 22952212</span></span>
<span><span class="co">## 2   chr1__DNAJC11_6634169__HMGCL_23801876  23801876       0 17188146</span></span>
<span><span class="co">## 3   chr1__TMEM201_9588910__STMN1_25900115  25900115       0 19114792</span></span>
<span><span class="co">## 4  chr1__SIKE1_114769483__CRTC2_153947675 153947675       0 39993799</span></span>
<span><span class="co">## 5  chr1__S100A3_153547328__NIT1_161118117 161118117       0 44603584</span></span>
<span><span class="co">## 6 chr1__ILF2_153661787__ADAMTS4_161189725 161189725       0  7628617</span></span></code></pre>
<p>Importantly, CNV inference adds the “CNV” partition:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span><span class="op">$</span><span class="va">partitions</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    global_expr CNV</span></span>
<span><span class="co">## V3           1   0</span></span>
<span><span class="co">## V4           1   0</span></span>
<span><span class="co">## V5           5   2</span></span>
<span><span class="co">## V6           1   0</span></span>
<span><span class="co">## V7           5   2</span></span>
<span><span class="co">## V8           5   1</span></span></code></pre>
<p>The calculation can be demanding. For example, it requires approximately 10 minutes on 2 cores (dual Intel(R) Xeon(R), 2.60GHz).</p>
</div>
<div class="section level3">
<h3 id="cnv-visualization">CNV visualization<a class="anchor" aria-label="anchor" href="#cnv-visualization"></a>
</h3>
<p>scMuffin provides two visualizations for CNVs: an heatmap and the cluster average profile plot. The heatmap is based on ComnplexHeatmap package and beside it, it is possible to visualize the genomic location of a series of given genes (specified in the argument <code>genes</code>), or, alternatively, the location of detected CNVs (argument <code>mark.detected.cnv = T</code>)</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">col_fun</span> <span class="op">&lt;-</span> <span class="fu">circlize</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.2</span>, <span class="fl">0.2</span>, length.out <span class="op">=</span> <span class="fl">11</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu">pals</span><span class="fu">::</span><span class="fu"><a href="https://kwstat.github.io/pals/reference/brewer.html" class="external-link">brewer.rdylbu</a></span><span class="op">(</span><span class="fl">11</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/heatmap_CNV.html">heatmap_CNV</a></span><span class="op">(</span>scMuffinList <span class="op">=</span> <span class="va">scMuffinList_demo</span>, genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"YBX1"</span>, <span class="st">"HNRNPM"</span><span class="op">)</span>, genes.labels <span class="op">=</span> <span class="cn">T</span>, col<span class="op">=</span><span class="va">col_fun</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/heatmap_CNV.html">heatmap_CNV</a></span><span class="op">(</span>scMuffinList <span class="op">=</span> <span class="va">scMuffinList_demo</span>, mark.detected.cnv <span class="op">=</span> <span class="cn">T</span>, col<span class="op">=</span><span class="va">col_fun</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_profile_CNV.html">plot_profile_CNV</a></span><span class="op">(</span>scMuffinList <span class="op">=</span> <span class="va">scMuffinList_demo</span>, cluster <span class="op">=</span> <span class="fl">0</span>, cex.points <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> </span></code></pre></div>
<p><img src="images/heatmap_CNV_PJ016_vignette_genes.jpg" width="600" alt="caption"><img src="images/heatmap_CNV_PJ016_vignette_cnv.jpg" width="600"></p>
<p>To function <code>plot_profile_CNV</code> plots the median CNV profile of a cluster:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_profile_CNV.html">plot_profile_CNV</a></span><span class="op">(</span>scMuffinList <span class="op">=</span> <span class="va">scMuffinList_demo</span>, cluster <span class="op">=</span> <span class="fl">0</span>, cex.points <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> </span></code></pre></div>
<p><img src="images/CNV_profile_ref_0.png_0.png" width="600"></p>
</div>
</div>
<div class="section level2">
<h2 id="transcriptional-complexity">Transcriptional complexity<a class="anchor" aria-label="anchor" href="#transcriptional-complexity"></a>
</h2>
<p>The transcriptional complexity can be quantified by means of the function <code>transcr_compl</code>. The corresponding summary element contains a <code>data.frame</code> with the number of cell transcripts, the number of genes detected in every cell, the TC-ratio (C), the TC-LMR (linear model residual) and the TC-H (entropy):</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMuffinList_demo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/transcr_compl.html">transcr_compl</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span><span class="op">$</span><span class="va">transcr_compl</span><span class="op">$</span><span class="va">summary</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    tot_counts n_genes        C        H          LM</span></span>
<span><span class="co">## V3      16233    1052 0.863711 6.523161 -0.02531052</span></span>
<span><span class="co">## V4      14286    1168 1.089642 6.736431  0.06163495</span></span>
<span><span class="co">## V5      15857    1217 1.022871 6.602594  0.04557971</span></span>
<span><span class="co">## V6      15157    1208 1.062197 6.763307  0.05702736</span></span>
<span><span class="co">## V7      15615    1172 1.000316 6.536441  0.03421426</span></span>
<span><span class="co">## V8      15286    1218 1.061952 6.685640  0.05785375</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="cell-proliferation-rate">Cell proliferation rate<a class="anchor" aria-label="anchor" href="#cell-proliferation-rate"></a>
</h2>
<p>Cell proliferation is quantified considering the maximum between the two scores of G1/S and G2/M gene sets:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMuffinList_demo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/proliferation_analysis.html">proliferation_analysis</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span><span class="op">)</span></span></code></pre></div>
<p>The proliferation score is stored in the <code>summary</code> element of <code>scMuffinList_demo$proliferation</code>:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span><span class="op">$</span><span class="va">proliferation</span><span class="op">$</span><span class="va">summary</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    Proliferation_score</span></span>
<span><span class="co">## V3          0.27013968</span></span>
<span><span class="co">## V4          0.42946960</span></span>
<span><span class="co">## V5         -0.17596045</span></span>
<span><span class="co">## V6          0.40310925</span></span>
<span><span class="co">## V7         -0.04090571</span></span>
<span><span class="co">## V8         -0.12102751</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="cell-state-trajectories">Cell state trajectories<a class="anchor" aria-label="anchor" href="#cell-state-trajectories"></a>
</h2>
<p>Diffusion maps identify differentiation trajectories. scMuffin relies on the diffusion pseudo time calculation available in the R package “destiny” <span class="citation">(Angerer et al. 2016)</span>. Here we calculate the diffusion map over the first 50 PC (to speed up computation), using a random cell to obtain diffusion pseudotimes:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMuffinList_demo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/diff_map.html">diff_map</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span>, root_cell <span class="op">=</span> <span class="st">"random"</span>, n_pcs<span class="op">=</span><span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p>A data.frame with the most important features of the analysis (the first two eigenvectors, pseudotime, branch information and whether a cell is a tip of the branch or not) are stored <code>scMuffinList_demo$diffusion_map_pseudo_t$summary</code>:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMuffinList_demo</span><span class="op">$</span><span class="va">diffusion_map_pseudo_t</span><span class="op">$</span><span class="va">summary</span></span></code></pre></div>
<pre><code><span><span class="co">##            DC1          DC2      dpt branch  tips</span></span>
<span><span class="co">## V3  0.02501839 -0.022810097 1.888927      2 FALSE</span></span>
<span><span class="co">## V4  0.03166680 -0.032411924 1.967455      2 FALSE</span></span>
<span><span class="co">## V5 -0.01895550 -0.010695236 1.675576      3 FALSE</span></span>
<span><span class="co">## V6  0.03301277 -0.039363092 1.990567      2 FALSE</span></span>
<span><span class="co">## V7 -0.01821403 -0.009274310 1.670158      3 FALSE</span></span>
<span><span class="co">## V8 -0.01831355 -0.009674605 1.671976      3 FALSE</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="cluster-enrichment-assessment">Cluster enrichment assessment<a class="anchor" aria-label="anchor" href="#cluster-enrichment-assessment"></a>
</h2>
<p>Cell clusters can be assessed for enrichment in quantitative and categorical values. The appropriate statistical test is automatically chosen according to feature type. In this example, we assess cluster enrichment for the feature <code>gene_set_scoring</code> (quantitative one) for the cluster defined by the partition <code>global_expr</code>:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMuffinList_demo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assess_cluster_enrichment.html">assess_cluster_enrichment</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span>, feature_name <span class="op">=</span> <span class="st">"gene_set_scoring"</span>, partition_id <span class="op">=</span> <span class="st">"global_expr"</span><span class="op">)</span></span></code></pre></div>
<p>In case of quantitative features the result is a list named GSEA, placed under the elements <code>cluster_data$global_expr</code> (partition id). The list contains the gene set table and the leading edge results. Here’s the result related to the gene set <code>CSEA_Angiogenesis</code>:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMuffinList_demo</span><span class="op">$</span><span class="va">cluster_data</span><span class="op">$</span><span class="va">global_expr</span><span class="op">$</span><span class="va">GSEA</span><span class="op">$</span><span class="va">gs_table</span><span class="op">$</span><span class="va">CSEA_Angiogenesis</span></span></code></pre></div>
<pre><code><span><span class="co">##   id         es p_val adj_p_val n_pos_perm n_neg_perm       nes        FDRq</span></span>
<span><span class="co">## 0  0 -0.1767187  0.01 0.0180000         98          2 0.0000000 1.000000000</span></span>
<span><span class="co">## 1  1  0.3124272  0.02 0.0300000        100          0 1.2511556 0.098737084</span></span>
<span><span class="co">## 2  2  0.1687780  1.00 1.0000000        100          0 0.6984846 0.974741676</span></span>
<span><span class="co">## 3  3 -0.1578852  0.01 0.0180000         99          1 0.0000000 1.000000000</span></span>
<span><span class="co">## 4  4 -0.2900841  0.01 0.0180000         97          3 0.0000000 1.000000000</span></span>
<span><span class="co">## 5  5  0.5501378  0.01 0.0180000         98          2 2.1523242 0.006888634</span></span>
<span><span class="co">## 6  6  0.3163922  0.10 0.1285714         97          3 1.2295078 0.106773823</span></span>
<span><span class="co">## 7  7  0.2876240  0.24 0.2700000         96          4 1.1208356 0.234213548</span></span>
<span><span class="co">## 8  8  0.4906238  0.01 0.0180000         86         14 1.7152713 0.006888634</span></span></code></pre>
<p>In this this example we assess cluster enrichment in relation to categorical feature, namely cell cycle phase:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMuffinList_demo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assess_cluster_enrichment.html">assess_cluster_enrichment</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span>, feature_name <span class="op">=</span> <span class="st">"CC_Phase"</span>, partition_id <span class="op">=</span> <span class="st">"global_expr"</span><span class="op">)</span></span></code></pre></div>
<p>As for categorical features the result is a list named ORA, placed under the elements <code>cluster_data$global_expr</code> (partition id). The list contains ORA results for every categorical value. Here’s the enrichment of clusters in terms of cells with value “G1”:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMuffinList_demo</span><span class="op">$</span><span class="va">cluster_data</span><span class="op">$</span><span class="va">global_expr</span><span class="op">$</span><span class="va">ORA</span><span class="op">$</span><span class="va">CC_Phase</span><span class="op">$</span><span class="va">G1</span></span></code></pre></div>
<pre><code><span><span class="co">##   id    N   wb  bb  bd wbd       exp        er            p        p_adj</span></span>
<span><span class="co">## 0  0 2828 2049 779 583 399 422.40700 0.9445866 9.931667e-01 1.000000e+00</span></span>
<span><span class="co">## 1  1 2828 2049 779 502   0 363.71924 0.0000000 1.000000e+00 1.000000e+00</span></span>
<span><span class="co">## 2  2 2828 2049 779 474 441 343.43211 1.2840966 3.156377e-34 1.420369e-33</span></span>
<span><span class="co">## 3  3 2828 2049 779 446 431 323.14498 1.3337667 5.121689e-47 4.609520e-46</span></span>
<span><span class="co">## 4  4 2828 2049 779 330 273 239.09830 1.1417898 2.642676e-06 5.946021e-06</span></span>
<span><span class="co">## 5  5 2828 2049 779 262 250 189.82956 1.3169709 1.192875e-23 3.578624e-23</span></span>
<span><span class="co">## 6  6 2828 2049 779 220 165 159.39887 1.0351391 2.124319e-01 3.823774e-01</span></span>
<span><span class="co">## 7  7 2828 2049 779 138   0  99.98656 0.0000000 1.000000e+00 1.000000e+00</span></span>
<span><span class="co">## 8  8 2828 2049 779 130  90  94.19024 0.9555130 8.275204e-01 1.000000e+00</span></span></code></pre>
<p>The enrichment analysis results appearing in these tables can be easily extracted and organized in a clusters-by-values table by means of <code>extract_cluster_enrichment_table</code>. For instance, here we extract GSEA NES and FDRq values, and ORA er (enrichment ratio) values:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gsea_fdr_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_cluster_enrichment_table.html">extract_cluster_enrichment_table</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span>,  partition_id <span class="op">=</span> <span class="st">"global_expr"</span>, type <span class="op">=</span> <span class="st">"GSEA"</span>, quantity <span class="op">=</span> <span class="st">"FDRq"</span><span class="op">)</span></span>
<span><span class="va">gsea_nes_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_cluster_enrichment_table.html">extract_cluster_enrichment_table</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span>,  partition_id <span class="op">=</span> <span class="st">"global_expr"</span>, type <span class="op">=</span> <span class="st">"GSEA"</span>, quantity <span class="op">=</span> <span class="st">"nes"</span><span class="op">)</span></span>
<span><span class="va">ora_p_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_cluster_enrichment_table.html">extract_cluster_enrichment_table</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span>,  partition_id <span class="op">=</span> <span class="st">"global_expr"</span>, type <span class="op">=</span> <span class="st">"ORA"</span>, quantity <span class="op">=</span> <span class="st">"p"</span><span class="op">)</span></span>
<span><span class="va">gsea_fdr_table</span></span></code></pre></div>
<pre><code><span><span class="co">##   CSEA_Angiogenesis CSEA_Apoptosis CSEA_Cell_Cycle CSEA_Differentiation</span></span>
<span><span class="co">## 0       1.000000000    0.091711957     1.000000000          1.000000000</span></span>
<span><span class="co">## 1       0.098737084    0.018292683     0.004608295          0.046718576</span></span>
<span><span class="co">## 2       0.974741676    0.008152174     1.000000000          0.266963293</span></span>
<span><span class="co">## 3       1.000000000    0.008152174     1.000000000          0.354467927</span></span>
<span><span class="co">## 4       1.000000000    0.831521739     1.000000000          0.997775306</span></span>
<span><span class="co">## 5       0.006888634    0.008152174     1.000000000          0.044493882</span></span>
<span><span class="co">## 6       0.106773823    0.018292683     0.231950845          0.845383760</span></span>
<span><span class="co">## 7       0.234213548    0.018292683     0.004608295          0.008898776</span></span>
<span><span class="co">## 8       0.006888634    0.813858696     1.000000000          0.008898776</span></span>
<span><span class="co">##   CSEA_DNA_damage CSEA_DNA_repair    CSEA_EMT CSEA_Hypoxia CSEA_Inflammation</span></span>
<span><span class="co">## 0     1.000000000     1.000000000 0.059325176  0.998883929       1.000000000</span></span>
<span><span class="co">## 1     0.306117909     0.005643341 1.000000000  1.000000000       1.000000000</span></span>
<span><span class="co">## 2     0.171301446     1.000000000 0.017797553  0.005580357       0.005586592</span></span>
<span><span class="co">## 3     1.000000000     1.000000000 1.000000000  0.005580357       0.013966480</span></span>
<span><span class="co">## 4     0.332220986     0.996613995 1.000000000  0.005580357       0.055865922</span></span>
<span><span class="co">## 5     0.979977753     1.000000000 0.008898776  0.005580357       0.101955307</span></span>
<span><span class="co">## 6     0.008898776     0.005643341 1.000000000  1.000000000       0.982122905</span></span>
<span><span class="co">## 7     0.038561364     0.005643341 1.000000000  1.000000000       1.000000000</span></span>
<span><span class="co">## 8     0.008898776     0.007054176 1.000000000  1.000000000       1.000000000</span></span>
<span><span class="co">##   CSEA_Invasion CSEA_Metastasis CSEA_Proliferation CSEA_Quiescence</span></span>
<span><span class="co">## 0   1.000000000     0.172991071        1.000000000     1.000000000</span></span>
<span><span class="co">## 1   0.008898776     1.000000000        0.005592841     1.000000000</span></span>
<span><span class="co">## 2   1.000000000     0.005580357        1.000000000     0.014833127</span></span>
<span><span class="co">## 3   1.000000000     0.005580357        1.000000000     1.000000000</span></span>
<span><span class="co">## 4   1.000000000     0.952008929        0.011185682     0.004944376</span></span>
<span><span class="co">## 5   1.000000000     0.005580357        1.000000000     0.627935723</span></span>
<span><span class="co">## 6   0.280311457     1.000000000        0.856823266     0.020408163</span></span>
<span><span class="co">## 7   0.008898776     1.000000000        0.005592841     1.000000000</span></span>
<span><span class="co">## 8   0.038561364     1.000000000        0.267058166     0.613102596</span></span>
<span><span class="co">##   CSEA_Stemness</span></span>
<span><span class="co">## 0    1.00000000</span></span>
<span><span class="co">## 1    0.00779510</span></span>
<span><span class="co">## 2    1.00000000</span></span>
<span><span class="co">## 3    1.00000000</span></span>
<span><span class="co">## 4    0.00935412</span></span>
<span><span class="co">## 5    1.00000000</span></span>
<span><span class="co">## 6    0.00779510</span></span>
<span><span class="co">## 7    0.00779510</span></span>
<span><span class="co">## 8    0.00779510</span></span></code></pre>
<p>These tables can be plotted with the function <code>plot_heatmap_features_by_clusters</code>. In the following example we plot NES values with a asterisks according to their significance:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_heatmap_features_by_clusters.html">plot_heatmap_features_by_clusters</a></span><span class="op">(</span>feature_source <span class="op">=</span> <span class="va">gsea_nes_table</span>, significance_matrix <span class="op">=</span> <span class="va">gsea_fdr_table</span>, sig_threshold <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/nes_fdr.png" width="300"></p>
<p>Similarly, it’s possible to extract the most significant “tags” of any clusters; for example, here we extract the top 3 tags by FDRq (GSEA) and p_adj (ORA):</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMuffinList_demo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_cluster_enrichment_tags.html">extract_cluster_enrichment_tags</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span>, partition_id <span class="op">=</span> <span class="st">"global_expr"</span>, n_max_per_cluster <span class="op">=</span> <span class="fl">3</span>, GSEA_selection_criterion<span class="op">=</span><span class="st">"FDRq"</span>, only_pos_nes <span class="op">=</span> <span class="cn">TRUE</span>, GSEA_selection_threshold <span class="op">=</span> <span class="fl">0.05</span>, ORA_selection_criterion <span class="op">=</span> <span class="st">"p_adj"</span>, ORA_selection_threshold <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span></code></pre></div>
<p>The results are placed under the <code>cluster_tags</code> element:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span><span class="op">$</span><span class="va">cluster_data</span><span class="op">$</span><span class="va">global_expr</span><span class="op">$</span><span class="va">cluster_tags</span><span class="op">$</span><span class="va">GSEA</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $`0`</span></span>
<span><span class="co">## [1] "CSEA_Apoptosis"  "CSEA_EMT"        "CSEA_Metastasis"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`1`</span></span>
<span><span class="co">## [1] "CSEA_DNA_repair" "CSEA_Cell_Cycle" "CSEA_Invasion"  </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`2`</span></span>
<span><span class="co">## [1] "CSEA_Metastasis"   "CSEA_Inflammation" "CSEA_Hypoxia"     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`3`</span></span>
<span><span class="co">## [1] "CSEA_Apoptosis"  "CSEA_Metastasis" "CSEA_Hypoxia"   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`4`</span></span>
<span><span class="co">## [1] "CSEA_Quiescence" "CSEA_Hypoxia"    "CSEA_Stemness"  </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`5`</span></span>
<span><span class="co">## [1] "CSEA_Metastasis"   "CSEA_Angiogenesis" "CSEA_Apoptosis"</span></span></code></pre>
<p>The results of cluster enrichment can be visualized by barplots and boxplots for, respectively, categorical values and quantitative values.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/barplot_cluster.html">barplot_cluster</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span>, partition_id <span class="op">=</span> <span class="st">"global_expr"</span>, feature_name <span class="op">=</span> <span class="st">"CC_Phase"</span>, feature_id <span class="op">=</span> <span class="st">"CC_Phase"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/boxplot_cluster.html">boxplot_cluster</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span>, feature_name <span class="op">=</span> <span class="st">"gene_set_scoring"</span>, partition_id <span class="op">=</span> <span class="st">"global_expr"</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/cluster_2_barplot.jpg" width="600" alt="caption"><img src="images/cluster_2_boxplot.jpg" width="600" alt="caption"></p>
</div>
<div class="section level2">
<h2 id="comparison-between-clusters">Comparison between clusters<a class="anchor" aria-label="anchor" href="#comparison-between-clusters"></a>
</h2>
<div class="section level3">
<h3 id="intra-dataset">Intra-dataset<a class="anchor" aria-label="anchor" href="#intra-dataset"></a>
</h3>
<p>The function <code>overlap_matrix</code> calculates the overlap coefficient between all-pairs of clusters of two or more partitions of the same cells (same dataset):</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scMuffinList_demo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/overlap_matrix.html">overlap_matrix</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span><span class="op">)</span></span></code></pre></div>
<p>The results are stored under the element `cluster_comparison```:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scMuffinList_demo</span><span class="op">$</span><span class="va">cluster_comparison</span><span class="op">$</span><span class="va">overlap_matrix</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                    CNV_0       CNV_1       CNV_2      CNV_3 global_expr_0</span></span>
<span><span class="co">## CNV_0         1.00000000 0.000000000 0.000000000 0.00000000    0.01372213</span></span>
<span><span class="co">## CNV_1         0.00000000 1.000000000 0.000000000 0.00000000    0.38250429</span></span>
<span><span class="co">## CNV_2         0.00000000 0.000000000 1.000000000 0.00000000    0.24528302</span></span>
<span><span class="co">## CNV_3         0.00000000 0.000000000 0.000000000 1.00000000    0.56948229</span></span>
<span><span class="co">## global_expr_0 0.01372213 0.382504288 0.245283019 0.56948229    1.00000000</span></span>
<span><span class="co">## global_expr_1 0.98007968 0.007968127 0.001992032 0.01362398    0.00000000</span></span>
<span><span class="co">##               global_expr_1 global_expr_2 global_expr_3 global_expr_4</span></span>
<span><span class="co">## CNV_0           0.980079681   0.002109705    0.00000000    0.23030303</span></span>
<span><span class="co">## CNV_1           0.007968127   0.373417722    0.36547085    0.57575758</span></span>
<span><span class="co">## CNV_2           0.001992032   0.504219409    0.55829596    0.07272727</span></span>
<span><span class="co">## CNV_3           0.013623978   0.155313351    0.09264305    0.12121212</span></span>
<span><span class="co">## global_expr_0   0.000000000   0.000000000    0.00000000    0.00000000</span></span>
<span><span class="co">## global_expr_1   1.000000000   0.000000000    0.00000000    0.00000000</span></span>
<span><span class="co">##               global_expr_5 global_expr_6 global_expr_7 global_expr_8</span></span>
<span><span class="co">## CNV_0           0.003816794   0.963636364   0.992753623   0.953846154</span></span>
<span><span class="co">## CNV_1           0.255725191   0.027272727   0.007246377   0.023076923</span></span>
<span><span class="co">## CNV_2           0.671755725   0.000000000   0.000000000   0.007692308</span></span>
<span><span class="co">## CNV_3           0.068702290   0.009090909   0.000000000   0.015384615</span></span>
<span><span class="co">## global_expr_0   0.000000000   0.000000000   0.000000000   0.000000000</span></span>
<span><span class="co">## global_expr_1   0.000000000   0.000000000   0.000000000   0.000000000</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="inter-dataset">Inter-dataset<a class="anchor" aria-label="anchor" href="#inter-dataset"></a>
</h3>
<p>Inter-dataset comparison is performed through the function <code>inter_dataset_comparison</code>, which require a list of Seurat objects and a gene set list. The activity of this gene set list will be assessed over the clusters of all datasets. In needed, the function <code>prepare_cluster_markers_list</code> provides the possibility to prepare a gene set list of cluster markers, starting</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataset_cmp_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/inter_dataset_comparison.html">inter_dataset_comparison</a></span><span class="op">(</span>seu_obj_list <span class="op">=</span> <span class="va">seu_obj_list</span>, gsl <span class="op">=</span> <span class="va">cluster_markers_list</span>, genes_max <span class="op">=</span> <span class="fl">500</span>, genes_min <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_heatmap_dataset_comparison.html">plot_heatmap_dataset_comparison</a></span><span class="op">(</span><span class="va">dataset_cmp_res</span>, outfile <span class="op">=</span> <span class="st">"vignette/heatmap_ds_cmp.png"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_heatmap_dataset_comparison.html">plot_heatmap_dataset_comparison</a></span><span class="op">(</span><span class="va">dataset_cmp_res</span>, outfile <span class="op">=</span> <span class="st">"vignette/heatmap_ds_cmp_p.png"</span>, type <span class="op">=</span> <span class="st">"significance"</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/heatmap_ds_cmp.png" width="600" alt="caption"><img src="images/heatmap_ds_cmp_p.png" width="600" alt="caption"></p>
<p>The function can be used to analyze any gene set. In the following example we assess the expression of CancerSEA gene sets across datasets:</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gsl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_gsls.html">prepare_gsls</a></span><span class="op">(</span>gs_sources <span class="op">=</span> <span class="st">"CancerSEA"</span>, genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">seu_obj_list</span>, <span class="va">rownames</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dataset_cmp_res_cancersea</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/inter_dataset_comparison.html">inter_dataset_comparison</a></span><span class="op">(</span>seu_obj_list <span class="op">=</span> <span class="va">seu_obj_list</span>, gsl <span class="op">=</span> <span class="va">gsl</span><span class="op">$</span><span class="va">CancerSEA</span>, genes_max <span class="op">=</span> <span class="fl">500</span>, genes_min <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_heatmap_dataset_comparison.html">plot_heatmap_dataset_comparison</a></span><span class="op">(</span><span class="va">dataset_cmp_res_cancersea</span>, outfile <span class="op">=</span> <span class="st">"vignette/heatmap_ds_cmp_cancer.png"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="images/heatmap_ds_cmp_cancer.png" width="600" alt=""><p class="caption">caption</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<!-- ## Comparison of partitions and definition of meta-clusters

Lastly, scMuffin provides some functions to compare 


```r
cl_list <- partitions_to_list(partitions = clust_obj)
ov_mat <- overlap_matrix(cl_list)
meta_cl <- meta_cluster(ov_mat, n_step = 10, max_clust = 15)
meta_cl_cell <- get_meta_clusters(cl_list, meta_cl)
plot_meta_clusters(ov_mat, meta_clusters = meta_cl)
```
-->
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Angerer2016" class="csl-entry">
Angerer, Philipp, Laleh Haghverdi, Maren Büttner, Fabian J. Theis, Carsten Marr, and Florian Buettner. 2016. <span>“<span class="nocase">destiny : diffusion maps for large-scale single-cell data in R</span>.”</span> <em>Bioinformatics</em> 32 (8): 1241–43. <a href="https://doi.org/10.1093/bioinformatics/btv715" class="external-link">https://doi.org/10.1093/bioinformatics/btv715</a>.
</div>
<div id="ref-Franzen2019" class="csl-entry">
Franzén, Oscar, Li-Ming Gan, and Johan L M Björkegren. 2019. <span>“<span class="nocase">PanglaoDB: a web server for exploration of mouse and human single-cell RNA sequencing data</span>.”</span> <em>Database</em> 2019 (January). <a href="https://doi.org/10.1093/database/baz046" class="external-link">https://doi.org/10.1093/database/baz046</a>.
</div>
<div id="ref-Hao2021" class="csl-entry">
Hao, Yuhan, Stephanie Hao, Erica Andersen-Nissen, William M. Mauck, Shiwei Zheng, Andrew Butler, Maddie J. Lee, et al. 2021. <span>“<span class="nocase">Integrated analysis of multimodal single-cell data</span>.”</span> <em>Cell</em> 184 (13): 3573–3587.e29. <a href="https://doi.org/10.1016/j.cell.2021.04.048" class="external-link">https://doi.org/10.1016/j.cell.2021.04.048</a>.
</div>
<div id="ref-Subramanian2005" class="csl-entry">
Subramanian, Aravind, Pablo Tamayo, Vamsi K. Mootha, Sayan Mukherjee, Benjamin L. Ebert, Michael A. Gillette, Amanda Paulovich, et al. 2005. <span>“<span class="nocase">Gene set enrichment analysis: A knowledge-based approach for interpreting genome-wide expression profiles</span>.”</span> <em>Proceedings of the National Academy of Sciences</em> 102 (43): 15545–50. <a href="https://doi.org/10.1073/pnas.0506580102" class="external-link">https://doi.org/10.1073/pnas.0506580102</a>.
</div>
<div id="ref-Yuan2019" class="csl-entry">
Yuan, Huating, Min Yan, Guanxiong Zhang, Wei Liu, Chunyu Deng, Gaoming Liao, Liwen Xu, et al. 2019. <span>“<span class="nocase">CancerSEA: a cancer single-cell state atlas</span>.”</span> <em>Nucleic Acids Research</em> 47 (D1): D900–908. <a href="https://doi.org/10.1093/nar/gky939" class="external-link">https://doi.org/10.1093/nar/gky939</a>.
</div>
<div id="ref-Yuan2018" class="csl-entry">
Yuan, Jinzhou, Hanna Mendes Levitin, Veronique Frattini, Erin C. Bush, Deborah M. Boyett, Jorge Samanamud, Michele Ceccarelli, et al. 2018. <span>“<span class="nocase">Single-cell transcriptome analysis of lineage diversity in high-grade glioma</span>.”</span> <em>Genome Medicine</em> 10 (1): 57. <a href="https://doi.org/10.1186/s13073-018-0567-9" class="external-link">https://doi.org/10.1186/s13073-018-0567-9</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Valentina Nale, Noemi Di Nanni, Alice Chiodi, Ettore Mosca.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
